### PR TITLE
Notification - Step 1

### LINKED ISSUE #145: Complete notifications system
Today Gitea can send e-mail for new Issues/PRs/Comments and that's it.

I think it should have a complete notifications system, like GitHub and GitLab.

- [x] Notifications would be saved in a database table
- [x] A page to manage them. Equivalent to https://github.com/notifications
- [x] Mark as read/unread, pinning a notification (visible even if read)
- [x] Subscribe/Unsubscribe button
- [x] Mark all notifications as read
- [ ] Mark all notifications as unread
- [ ] Automatic subscribe on Open/Comment, etc
- [ ] Also more refined watch options?
- [ ] E-mail read means notification gone. Each user should have the ability to disable this
- [ ] E-Mail notifications for pushes

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/39150685-complete-notifications-system?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by andreynering**
First step to resolve #145

This deprecates #429

I plan 2 steps:

- [x] Create database table and create/update notifications (this one)
- [ ] Actually have UI for notifications

### CODE REVIEW

File: models/notification.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
+package models
```
Comment (lunny) [inline models/notification.go:5]: copyright header

File: models/notification.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+	Status NotificationStatus `xorm:"SMALLINT INDEX NOT NULL"`
+	Source NotificationSource `xorm:"SMALLINT INDEX NOT NULL"`
+
+	IssueID  int64  `xorm:"INDEX NOT NULL"`
+	CommitID string `xorm:"INDEX"`
+
+	UpdatedBy int64 `xorm:"INDEX NOT NULL"`
+
+	Issue       *Issue       `xorm:"-"`
```
Comment (lunny) [inline models/notification.go]: `Repo *Repository `xorm:"="` is convenience.

File: models/notification.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+		Where("user_id = ?", user.ID).
+		And("status = ?", status).
+		OrderBy("updated_unix DESC").
+		Find(&notifications)
+	return
+}
+
+// GetRepo returns the repo of the notification
+func (n *Notification) GetRepo() (repo *Repository, err error) {
+	repo = new(Repository)
```
Comment (lunny) [inline models/notification.go]: Store repo to struct field for cache.

File: models/notification.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+	_, err = x.
+		Where("id = ?", n.RepoID).
+		Get(repo)
+	return
+}
+
+// GetIssue returns the issue of the notification
+func (n *Notification) GetIssue() (issue *Issue, err error) {
+	issue = new(Issue)
```
Comment (lunny) [inline models/notification.go]: Store issue to struct field for cache.

File: modules/notification/notification.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
+package notification
```
Comment (lunny) [inline modules/notification/notification.go:5]: copyright header

**Review by lunny — CHANGES_REQUESTED** [review]
otherwise LGTM

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
LGTM

**Comment by bkcsoft** [comment]
LGTM :)

**Pull request merged**